### PR TITLE
Add function arrayReduceInRanges

### PR DESCRIPTION
**Pull request by hczhcz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add function `arrayReduceInRanges` which aggregates array elements in given ranges.

Detailed description / Documentation draft:

>Applies an aggregate function to array elements in given ranges and returns an array containing the result corresponding to each range. The function will return the same result as multiple `arrayReduce(agg_func, arraySlice(arr1, index, length), ...)`.
>
>Example:
>
>```sql
>SELECT arrayReduceInRanges(
>    'sum',
>    [1, 2],
>    [2, 3],
>    [100, 20, 3000, 4]
>)
>```
>
>```text
>┌─arrayReduceInRanges('sum', [1, 2], [2, 3], [100, 20, 3000, 4])─┐
>│ [120,3024]                                                     │
>└────────────────────────────────────────────────────────────────┘
>```

### CODE REVIEW

**Comment by amosbird** [comment]
The example is a bit misleading, as I interpret it as supplying two ranges : `[1, 2]` and `[2, 3]` at first sight.

**Comment by hczhcz** [comment]
This one might be better.

```sql
SELECT arrayReduceInRanges(
    'sum',
    [1, 2, 3, 4],
    [5, 3, 4, 4],
    [1000000, 200000, 30000, 4000, 500, 60, 7]
) AS res
```

```text
┌─res─────────────────────────┐
│ [1234500,234000,34560,4567] │
└─────────────────────────────┘
```

**Comment by qoega** [comment]
Will it be more clear if you use array of tuples?
```sql
SELECT arrayReduceInRanges(
    'sum',
    [(1, 5), (2, 3), (3, 4), (4, 4)],
    [1000000, 200000, 30000, 4000, 500, 60, 7]
) AS res
```

If you have a case, when index and length are in separate columns you can just use arrayZip
```sql
SELECT
    [1, 2, 3, 4, 5] AS x,
    [1, 2, 3, 4, 5] AS y,
    arrayZip(x, y) AS z

┌─x───────────┬─y───────────┬─z───────────────────────────────┐
│ [1,2,3,4,5] │ [1,2,3,4,5] │ [(1,1),(2,2),(3,3),(4,4),(5,5)] │
└─────────────┴─────────────┴─────────────────────────────────┘
```

File: docs/en/query_language/functions/array_functions.md [hunk 1]
```diff
@@ -814,6 +814,27 @@ SELECT arrayReduce('uniqUpTo(3)', [1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
...
+    [1000000, 200000, 30000, 4000, 500, 60, 7]
+) AS res
+```
+
+```text
+┌─res─────────────────────────┐
+│ [1234500,234500,34560,4567] │
+└─────────────────────────────┘
+```
+
```
Comment (qoega) [inline docs/en/query_language/functions/array_functions.md]: It is better to explicitly describe parameters.
By the way there is a function doc template:
https://github.com/ClickHouse/ClickHouse/blob/master/docs/dscr-templates/template-function.md

**Comment by alexey-milovidov** [comment]
Sorry, it looks like this PR was not merged with upstream properly. Could you please rebase your commits?

File: dbms/src/Functions/array/arrayReduceInRanges.cpp [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+            size_t place_end = local_end / minimum_step;
+
+            AggregateDataPtr place = arena->alignedAlloc(agg_func.sizeOfData(), agg_func.alignOfData());
+            try
+            {
+                agg_func.create(place);
+            }
+            catch (...)
+            {
+                agg_func.destroy(place);
```
Comment (alexey-milovidov) [inline dbms/src/Functions/array/arrayReduceInRanges.cpp]: This is wrong. If constructor has thrown an exception, destructor must not be called.

**Review by alexey-milovidov — APPROVED** [review]
I will merge to read the algorithm more carefully (it's difficult to understand).

**Pull request merged**